rest-lib/src/main/scala/com/gu/mediaservice/lib/management/InnerServiceStatusCheckController.scala

```diff
@@ -0,0 +1,58 @@
+package com.gu.mediaservice.lib.management
+
+import com.gu.mediaservice.lib.argo.ArgoHelpers
+import com.gu.mediaservice.lib.auth.Authentication
+import com.gu.mediaservice.lib.auth.Authentication.InnerServicePrincipal
+import com.gu.mediaservice.lib.config.Services
+import play.api.libs.json.{JsString, JsValue, Json, Writes}
+import play.api.libs.ws.{WSClient, WSRequest}
+import play.api.mvc.{BaseController, ControllerComponents}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+case class WhoAmIResponse (baseUri: String, status: Int, body: JsValue)
+object WhoAmIResponse { implicit val writes: Writes[WhoAmIResponse] = Json.writes[WhoAmIResponse] }
+
+class InnerServiceStatusCheckController(
+  auth: Authentication,
+  override val controllerComponents: ControllerComponents,
+  services: Services,
+  ws: WSClient
+)(implicit ec: ExecutionContext)
+  extends BaseController with ArgoHelpers {
+
+  private def safeJsonParse(maybeJsonStr: String) = Try(Json.parse(maybeJsonStr)).getOrElse(JsString(maybeJsonStr))
+
+  private def callAllInternalServices(authenticator: WSRequest => WSRequest) = {
+    val whoAmIFutures = services.allInternalUris.map { baseUri =>
+        authenticator(ws.url(s"$baseUri/management/whoAmI")).get
+          .map(resp => WhoAmIResponse(baseUri, resp.status, safeJsonParse(resp.body)))
+          .recover{
+            case throwable: Throwable => WhoAmIResponse(baseUri, SERVICE_UNAVAILABLE, Json.obj(
+            "errorMessage" -> throwable.getMessage,
+            "stackTrace" -> throwable.getStackTrace.map(_.toString)
+          ))}
+    }
+
+    Future.sequence(whoAmIFutures).map { whoAmIResponses =>
+      val overallStatus = whoAmIResponses.map(_.status).max
+      new Status(overallStatus)(Json.toJson(whoAmIResponses.map(resp => resp.baseUri -> resp).toMap))
+    }
+  }
+
+  def whoAmI = auth.async { request =>
+    request.user match {
+      case principal: InnerServicePrincipal if !principal.identity.contains(" via ") =>
```

Comment:
Worth parameterising? The only reason I mention – IIUC if this doesn't match (e.g. a dev makes a change to 'via' as it's added) we'll get stuck in an exponentiating loop.

Reply:
good point, funnily enough thats how i first implemented it, then chose the above for succinctness (but on reflection not worth the risk of it going haywire). reimplemented `depth` param in https://github.com/guardian/grid/pull/3342/commits/f1ff6ea5ba88962bb8763bd538e224d091dc6197 (capped at max 3)